src/dbnode/storage/series/buffer.go

```diff
@@ -183,78 +240,138 @@ func (b *dbBuffer) Write(
 	annotation []byte,
 ) error {
 	now := b.nowFn()
-	futureLimit := now.Add(1 * b.bufferFuture)
-	pastLimit := now.Add(-1 * b.bufferPast)
-	if !futureLimit.After(timestamp) {
-		return m3dberrors.ErrTooFuture
+	bucketStart := timestamp.Truncate(b.blockSize)
+	key := b.writableBucketKey(timestamp)
+
+	if b.isRealTime(timestamp) {
+		idx := b.writableBucketIdx(timestamp)
+
+		if b.bucketsRealTime[idx].needsReset(now, bucketStart) {
+			b.DrainAndReset()
+		}
+		// If there was previously a non-realtime metric written to this bucket
+		// already, then use that bucket.
+		if _, ok := b.bucketsNotRealTime[key]; ok {
+			b.makeBucketRealTime(key, idx)
+		}
+
+		return b.bucketsRealTime[idx].write(now, timestamp, value, unit, annotation)
 	}
-	if !pastLimit.Before(timestamp) {
-		return m3dberrors.ErrTooPast
+
+	if !b.opts.RetentionOptions().AnyWriteTimeEnabled() {
+		return errAnyWriteTimeNotEnabled
 	}
 
-	bucketStart := timestamp.Truncate(b.blockSize)
-	idx := b.writableBucketIdx(timestamp)
-	if b.buckets[idx].needsReset(bucketStart) {
-		// Needs reset
-		b.DrainAndReset()
+	if _, ok := b.bucketsNotRealTime[key]; !ok {
+		b.initializeBucket(key)
 	}
 
-	return b.buckets[idx].write(timestamp, value, unit, annotation)
+	if b.bucketsNotRealTime[key].needsDrain(now, bucketStart) {
+		b.bucketDrain(now, bucketIDNotRealtime(key), bucketStart)
+
+		if b.bucketsNotRealTime[key].isStale(now) {
+			b.removeBucket(key)
+		}
+	}
+
+	return b.bucketsNotRealTime[key].write(now, timestamp, value, unit, annotation)
```

Comment:
Seems like this would panic if you just removed the bucket you're trying to write to

Reply:
If I'm right about that, make sure to add a test case

Reply:
This is a good catch and was a bug. It makes no sense to remove a bucket due to staleness right before a write, since that will cause the bucket to be removed and then reinitialized right after. The only place where it makes sense to remove a stale bucket is during a bucketTick - I've cleaned this code up.